File: sdk/python/module/helpers/mqtt_client.py
import os
import collections
import time
import json
import paho.mqtt.client as mqtt
import ssl
import Queue

import sdk.python.constants as constants
import sdk.python.utils.exceptions as exception
from sdk.python.module.helpers.message import Message
from sdk.python.module.helpers.mqtt_consumer import Mqtt_consumer

class Mqtt_client():
    def __init__(self, module):
        # we need the module's object to call its methods
        self.module = module
        # mqtt object
        self.gateway = None
        # track the topics subscribed
        self.topics_to_subscribe = []
        self.topics_subscribed = []
        self.topics_to_wait = []
        # queue messages while offline
        self.publish_queue = collections.deque(maxlen=300)
        # queue configuration messages while not configured
        self.configuration_queue = collections.deque(maxlen=500)
        # use a queue for exchanging incoming messages with consumer threads
        self.consumer_queue = Queue.Queue(maxsize=0)
        # number of consumer threads to create
        self.consumer_threads = 1
        # initialize consumer threads
        self.consumers = []
        for i in range(0, self.consumer_threads):
            self.consumers.append(Mqtt_consumer(i, self))
        
    # connect to the MQTT broker
    def __connect(self):
        # setup TLS for tcp transport if needed
        if self.module.gateway_ssl and self.module.gateway_transport == "tcp": 
            self.gateway.tls_set(ca_certs=self.module.gateway_ca_cert, certfile=self.module.gateway_certfile, keyfile=self.module.gateway_keyfile)
            # do not check for certificate validity
            self.gateway.tls_insecure_set(True)
        # setup SSL for websocket transport if needed
        elif self.module.gateway_ssl and self.module.gateway_transport == "websockets":
            self.gateway.tls_set(cert_reqs=ssl.CERT_NONE)
        # try to connect to the gateway until succeeding
        while self.module.connected == False:
            try:
                self.module.log_debug("Connecting to "+self.module.gateway_hostname+":"+str(self.module.gateway_port)+" ("+self.module.gateway_transport+", ssl="+str(self.module.gateway_ssl)+")")
                self.gateway.connect(self.module.gateway_hostname, self.module.gateway_port)
                # TODO: this has to be moved into on_connect but is never called if moved
                self.module.connected = True
                # TODO: last will? e.g. log a disconnecting message
            except Exception,e:
                self.module.log_warning("Unable to connect to "+self.module.gateway_hostname+":"+str(self.module.gateway_port)+" - "+exception.get(e))
                self.module.sleep(10)

    # subscribe to a given topic
    def __subscribe(self, topic):
        self.module.log_debug("Subscribing topic "+topic)
        self.gateway.unsubscribe(topic)
        self.gateway.subscribe(topic, qos=2)
        
    # Build the full topic (e.g. egeoffrey/v1/<house_id>/<from_module>/<to_module>/<command>/<args>)
    def __build_topic(self, house_id, from_module, to_module, command, args):
        if args == "": args = "null"
        return "/".join(["egeoffrey", constants.API_VERSION, house_id, from_module, to_module, command, args])

    # publish a given topic 
    def publish(self, house_id, to_module, command, args, payload_data, retain=False):
        # serialize the payload in a json format
        payload = payload_data
        if payload is not None: payload = json.dumps(payload)
        # build the topic to publish to
        topic = self.__build_topic(house_id, self.module.fullname, to_module, command, args)
        # publish if connected
        if self.module.connected:
            info = self.gateway.publish(topic, payload, retain=retain, qos=2)
        # queue the message if offline
        else:
            self.publish_queue.append([topic, payload, retain])
            
    # unsubscribe from a topic
    def unsubscribe(self, topic):
        if topic not in self.topics_subscribed: return
        self.module.log_debug("Unsubscribing from "+topic)
        self.topics_subscribed.remove(topic)
        self.gateway.unsubscribe(topic)
    
    # called from module. Connect to the MQTT broker and subscribe to the requested topics
    def start(self):
        # set client id. Format: egeoffrey-<house_id>-<scope>-<name>
        self.__client_id = "-".join(["egeoffrey", self.module.house_id, self.module.scope, self.module.name])
        # get an instance of the MQTT client object
        if self.module.persistent_client: self.module.log_debug("Configuring as persistent mqtt client")
        clean_session = False if self.module.persistent_client else True
        self.gateway = mqtt.Client(client_id=self.__client_id, clean_session=clean_session, userdata=None, transport=self.module.gateway_transport)
        # define what to do upon connect
        def __on_connect(client, userdata, flags, rc):
            try:
                if rc == 0:
                    self.module.log_debug("Connected to "+self.module.gateway_hostname+":"+str(self.module.gateway_port))
                    # call user's callback
                    self.module.on_connect()
                    # subscribe to the requested topics
                    for topic in self.topics_to_subscribe: 
                        self.__subscribe(topic)
                        self.topics_subscribed.append(topic)
                    # if there are message in the queue, send them
                    while True:
                        try:
                            entry = self.publish_queue.popleft()
                            self.gateway.publish(entry[0], entry[1], retain=entry[2], qos=2)
                        except IndexError:
                            break
                else:
                    # unable to connect, retry
                    self.module.log_error("Cannot connect: " + mqtt.connack_string(rc))
                    self.module.connected = False
                    self.__connect()
            except Exception,e:
                self.module.log_error("runtime error in __on_connect(): "+exception.get(e))
            
        # what to do when receiving a message
        def __on_message(client, userdata, msg):
            try:
                # parse the incoming request into a message data structure
                message = Message()
                message.parse(msg.topic, msg.payload, msg.retain)
                if self.module.verbose: self.module.log_debug("Received message "+message.dump(), False)
            except Exception,e:
                self.module.log_error("Invalid message received on "+msg.topic+" - "+msg.payload+": "+exception.get(e))
                return
            # ensure this message is for this house
            if message.house_id != "*" and message.house_id != self.module.house_id:
                self.module.log_debug("received message for the wrong house "+message.house_id+": "+message.dump())
                return
            # queue the message
            try:
                queue_size = self.consumer_queue.qsize()
                # print a warning if the incoming queue is getting too big
                if queue_size > 100:
                    self.module.log_warning("the incoming message queue is getting too big ("+str(queue_size)+" messages)")
                # if really too big, there is something wrong happening, ask the watchdog to restart our module
                if queue_size > 500 and self.module.watchdog is not None:
                    self.module.log_error("the incoming message queue is too big, requesting our watchdog to restart the module")
                    self.module.watchdog.restart_module(self.module.fullname)
                    return
                # queue the message
                self.consumer_queue.put_nowait(message)
            except Exception,e:
                self.module.log_error("Unable to queue incoming message: "+exception.get(e))

        # what to do upon disconnect
        def __on_disconnect(client, userdata, rc):
            self.module.connected = False
            # call user's callback
            try: 
                self.module.on_disconnect()
            except Exception,e: 
                self.module.log_error("runtime error during on_disconnect(): "+exception.get(e))
            if rc == 0:
                self.module.log_debug("Disconnected from "+self.module.gateway_hostname+":"+str(self.module.gateway_port))
            else:
                self.module.log_warning("Unexpected disconnection, reconnecting...")
                self.__connect()
            
        # set callbacks for mqtt
        self.gateway.on_connect = __on_connect
        self.gateway.on_message = __on_message
        self.gateway.on_disconnect = __on_disconnect
        # connect to the gateway
        self.gateway.username_pw_set(self.module.house_id, password=self.module.house_passcode)
        self.__connect()
        # start loop 
        try: 
            # start message consumer threads
            for consumer in self.consumers:
                consumer.start()
            # start mqtt network thread
            self.gateway.loop_start()
        except Exception,e: 
            self.module.log_error("Unexpected runtime error: "+exception.get(e))

    # add a listener for the given request
    def add_listener(self, from_module, to_module, command, args, wait_for_it):
        topic = self.__build_topic("+", from_module, to_module, command, args)
        if wait_for_it:
            # if this is mandatory topic, unconfigure the module and add it to the list of topics to wait for
            self.topics_to_wait.append(topic)
            self.module.configured = False
            self.module.log_debug("will wait for configuration on "+topic)
        # if connected, subscribe the topic and keep track of it
        if self.module.connected: 
            if topic in self.topics_subscribed: return topic
            self.__subscribe(topic)
            self.topics_subscribed.append(topic)
        # if not connected, will subscribe once connected
        else:
            if topic in self.topics_to_subscribe: return topic
            self.topics_to_subscribe.append(topic)
        # return the topic so the user can unsubscribe from it if needed
        return topic
            
    # disconnect from the MQTT broker
    def stop(self):
        # stop all message consumer threads
        for consumer in self.consumers:
            consumer.join()
        # do nothing if not connected to the gateway
        if self.gateway == None: return
        # stop the mqtt network thread
        self.gateway.loop_stop()
        # disconnect from the gateway
        self.gateway.disconnect()
        try:
            self.module.on_disconnect()
        except Exception,e: 
            self.module.log_error("runtime error during on_disconnect(): "+exception.get(e))

File: javascript/module/helpers/mqtt_client.js


class Mqtt_client {
    constructor(module) {
        // TODO: certs
        // we need the module's object to call its methods
        this.__module = module
        // mqtt object
        this.__gateway = null;
        // track the topics subscribed
        this.__topics_to_subscribe = []
        this.__topics_subscribed = []
        this.__topics_to_wait = []
        // queue messages while offline
        this.__queue = []
    }

    // connect to the MQTT broker
    __connect() {
        var this_class = this
        var __on_connect = function(reconnect, url) {
            this_class.__module.log_info("Connected to eGeoffrey gateway "+this_class.__module.gateway_hostname+":"+this_class.__module.gateway_port)
            this_class.__module.connected = true
            this_class.__module.on_connect()
            // subscribe to the requested topics
            for (var topic of this_class.__topics_to_subscribe) {
                this_class.__subscribe(topic)
                this_class.__topics_subscribed.push(topic)
            }
            // there are message in the queue, send them
            if (this_class.__queue.length > 0) { 
                for (var entry of this_class.__queue) {
                    this_class.__gateway.send(entry[0], entry[1], 0, entry[2])
                }
                this_class.__queue = []
            }
        }
        
        // connect to the gateway
        var __on_failure = function() {
            this_class.__module.log_error("Unable to connect to "+this_class.__module.gateway_hostname+":"+this_class.__module.gateway_port)
        }
        try {
            this.__module.log_debug("Connecting to "+this.__module.gateway_hostname+":"+this.__module.gateway_port+" (ssl="+this.__module.gateway_ssl+")")
            var connect_options = {
                "userName": this.__module.house_id,
                "password": this.__module.house_passcode,
                "onSuccess": __on_connect, 
                "onFailure": __on_failure, 
                "timeout": 2, 
                "useSSL": this.__module.gateway_ssl
            }
            this.__gateway.connect(connect_options)
        } catch(e) {
            this.__module.log_error("Unable to connect to "+this.__module.gateway_hostname+":"+this.__module.gateway_port+" "+get_exception(e))
        }
    }
    
    // subscribe to a given topic
    __subscribe(topic, qos=0) {
        this.__module.log_debug("Subscribing topic "+topic)
        this.__gateway.subscribe(topic, {"qos": qos})
    }
    
    // Build the full topic (e.g. egeoffrey/v1/<house_id>/<from_module>/<to_module>/<command>/<args>)
    __build_topic(house_id, from_module, to_module, command, args) {
        if (args == "") args = "null"
        return ["egeoffrey", constants["API_VERSION"], house_id, from_module, to_module, command, args].join("/")
    }
    
    // publish payload to a given topic (queue the message while offline)
    publish(house_id, to_module, command, args, payload_data, retain=false) {
        var payload = payload_data
        if (payload != null) payload = JSON.stringify(payload)
        else {
            var buffer = new ArrayBuffer(1)
            buffer[0] = null
            payload = buffer
        }
        var topic = this.__build_topic(house_id, this.__module.fullname, to_module, command, args)
        if (this.__module.connected) this.__gateway.send(topic, payload, 0, retain=retain)
        else this.__queue.push([topic, payload, retain])
    }
    
    // unsubscribe from a topic
    unsubscribe(topic) {
        this.__module.log_debug("Unsubscribing from "+topic)
        this.__topics_subscribed.remove(topic)
        this.__gateway.unsubscribe(topic)
    }
    
    // connect to the MQTT broker and subscribed to the requested topics
    start() {
        // set client id . Format: egeoffrey-<house_id>-<scope>-<name>
        this.__client_id = ["egeoffrey", this.__module.house_id, this.__module.scope, this.__module.name].join("-")
        // get an instance of the MQTT client
        this.__gateway = new Paho.MQTT.Client(this.__module.gateway_hostname, Number(this.__module.gateway_port), this.__client_id);
        // define what to do upon connect
        var this_class = this
        var __on_connect = function(reconnect, url) {
            this_class.__module.log_debug("Connected to "+this_class.__module.gateway_hostname+":"+this_class.__module.gateway_port)
            this_class.__module.connected = true
            // subscribe to the requested topics
            for (var topic of this_class.__topics_to_subscribe) {
                this_class.__subscribe(topic)
                this_class.__topics_subscribed.push(topic)
            }
        }

        // what to do when receiving a message
        var __on_message = function(msg) {
            if (msg == null) return
            try {
                // parse the incoming request into a message data structure
                var message = new Message()
                message.parse(msg.destinationName, msg.payloadString, msg.retained)
                if (this_class.__module.verbose) this_class.__module.log_debug("Received message "+message.dump(), false)
            } catch (e) {
                this_class.__module.log_error("Invalid message received on "+msg.destinationName+" - "+msg.payloadString+": "+get_exception(e))
                return
            }
            // ensure this message is for this house
            if (message.house_id != "*" && message.house_id != this_class.__module.house_id) {
                this_class.__module.log_warning("received message for the wrong house "+message.house_id+": "+message.dump())
                return
            }
            try {
                // identify the subscribed topic which caused this message to get here
                for (var pattern of this_class.__topics_subscribed) {
                    if (topic_matches_sub(pattern, message.topic)) {
                        // if the message is a configuration
                        if (message.sender == "controller/config" && message.command == "CONF") {
                            // notify the module about the configuration just received
                            try {
                                this_class.__module.on_configuration(message)
                            } catch(e) {
                                this_class.__module.log_error("runtime error during on_configuration() - "+message.dump()+": "+get_exception(e))
                                return
                            }
                            // check if we had to wait for this to start the module
                            if (this_class.__topics_to_wait.length > 0) {
                                for (var req_pattern of this_class.__topics_to_wait) {
                                    if (topic_matches_sub(req_pattern, message.topic)) {
                                        this_class.__module.log_debug("received configuration "+message.topic)
                                        var index = this_class.__topics_to_wait.indexOf(req_pattern)
                                        this_class.__topics_to_wait.splice(index, 1)
                                        // if there are no more topics to wait for, start the module
                                        if (this_class.__topics_to_wait.length == 0) { 
                                            this_class.__module.log_debug("Configuration completed for "+this_class.__module.fullname+", starting the module...")
                                            this_class.__module.configured = true
                                            try { 
                                                this_class.__module.on_start()
                                            } catch(e) {
                                                this_class.__module.log_error("runtime error during on_start(): "+get_exception(e))
                                            }
                                        } else {
                                            this_class.__module.log_debug("still waiting for configuration on "+JSON.stringify(this_class.__topics_to_wait))
                                        }
                                    }
                                }
                            }
                        // handle internal messages
                        } else if (message.command == "PING") {
                            message.reply()
                            message.command = "PONG"
                            this_class.__module.send(message)
                        // notify the module about this message (only if fully configured)
                        } else {
                            if (this_class.__module.configured) {
                                try {
                                    this_class.__module.on_message(message)
                                } catch(e) {
                                    this_class.__module.log_error("runtime error during on_message(): "+get_exception(e))
                                }
                            }
                        }
                        // avoid delivering the same message multiple times for overlapping subscribers
                        return 
                    }
                }
            } catch(e) {
                this_class.__module.log_error("Cannot handle request: "+get_exception(e))
            }
        }

        // what to do upon disconnect
        var __on_disconnect = function(response) {
            this_class.__module.connected = false
            if (response.errorCode == 0) this_class.__module.log_debug("Disconnected from "+this_class.__module.gateway_hostname+":"+this_class.__module.gateway_port)
            else this_class.__module.log_warning("Unexpected disconnection")
            try {
                this_class.__module.on_disconnect()
            } catch(e) {
                this_class.__module.log_error("runtime error during on_disconnect(): "+get_exception(e))
            }
        }
        
        // configure callbacks
        this.__gateway.onMessageArrived = __on_message
        this.__gateway.onConnectionLost = __on_disconnect
        // connect to the gateway
        this.__connect({"userName": this.__module.house_id, "password": this.__module.house_passcode})
    }
    
    // add a listener for the given request
    add_listener(from_module, to_module, command, filter, wait_for_it) {
        var topic = this.__build_topic("+", from_module, to_module, command, filter)
        if (wait_for_it) {
            // if this is mandatory topic, unconfigure the module and add it to the list of topics to wait for
            if (wait_for_it) {
                this.__topics_to_wait.push(topic)
                this.__module.configured = false
                this.__module.log_debug("will wait for configuration on "+topic)
            }
        } 
        // subscribe the topic and keep track of it
        if (this.__module.connected) {
            if (this.__topics_subscribed.includes(topic)) return topic
            this.__subscribe(topic)
            this.__topics_subscribed.push(topic)
        }
        // if not connected, will subscribed once connected
        else {
            if (this.__topics_to_subscribe.includes(topic)) return topic
            this.__topics_to_subscribe.push(topic)
        }
        return topic
    }
    
    // disconnect from the MQTT broker
    stop() {
        if (this.__gateway == null) return
        try {
            if (this.__gateway.isConnected()) {
                this.__gateway.disconnect()
                try {
                    this.__module.on_disconnect()
                } catch(e) {
                    this_class.__module.log_error("runtime error during on_disconnect(): "+get_exception(e))
                }
            }
            this.__module.connected = false
        } catch(e) {
            this.__module.log_error("Unable to disconnect from "+this.__module.gateway_hostname+":"+this.__module.gateway_port+" "+get_exception(e))
        }
    }
}